File: samples/java/app/src/main/java/org/remdev/wetweexapp/generator/GeneratorCoordinator.java
package org.remdev.wetweexapp.generator;

import androidx.annotation.NonNull;
import org.remdev.wetweex.component.ProgressWatchingCoordinator;
import org.remdev.wetweex.execution.ExecutableTask;
import org.remdev.wetweexapp.Strategy;
import org.remdev.wetweexapp.domain.services.TextGenerator;

import java.util.concurrent.Callable;

public class GeneratorCoordinator extends ProgressWatchingCoordinator<GeneratorViewModel> {

    private final TextGenerator textGenerator;

    public GeneratorCoordinator(@NonNull GeneratorViewModel viewModel, TextGenerator textGenerator) {
        super(viewModel);
        this.textGenerator = textGenerator;
    }

    public void generateString(Strategy strategy, Long millis) {
        final Callable<String> generationAction = () -> textGenerator.generateText(millis).getText();
        ExecutableTask<String> task;
        switch(strategy) {
            case NON_CANCELLABLE:
                task = ExecutableTask.nonCancellable(generationAction);
                break;
            case CRUCIAL:
                task = ExecutableTask.crucial(generationAction);
                break;
            case CANCELLABLE:
            default:
                task = ExecutableTask.cancellable(generationAction);
                break;
        }

        task.onSuccess(it -> getViewModel().getGeneratedText().setOrPost(it))
                .onCancel(() -> getViewModel().getGeneratedText().setOrPost("Cancelled"))
                .onError(it -> getViewModel().getGeneratedText().setOrPost(it.getMessage()));

        submitTask("Generate string: " + strategy, task);
    }
}

File: samples/kotlin/app/src/main/java/org/remdev/wetweexapp/generator/GeneratorCoordinator.kt
package org.remdev.wetweexapp.generator

import org.remdev.wetweex.component.ProgressWatchingCoordinator
import org.remdev.wetweex.execution.ExecutableTask
import org.remdev.wetweexapp.Strategy
import org.remdev.wetweexapp.domain.services.TextGenerator

class GeneratorCoordinator(generatorViewModel: GeneratorViewModel, private val textGenerator: TextGenerator)
    : ProgressWatchingCoordinator<GeneratorViewModel>(generatorViewModel) {

    fun generateString(strategy: Strategy, millis: Long) {
        val generationAction = {
            textGenerator.generateText(millis).text
        }
        val task = when(strategy) {
            Strategy.CANCELLABLE -> ExecutableTask.cancellable(generationAction)
            Strategy.NON_CANCELLABLE -> ExecutableTask.nonCancellable(generationAction)
            Strategy.CRUCIAL -> ExecutableTask.crucial(generationAction)
        }.onSuccess {
            viewModel.generatedText.setOrPost(it)
        }.onCancel {
            viewModel.generatedText.setOrPost("Cancelled")
        }.onError {
            viewModel.generatedText.setOrPost(it.message)
        }

        submitTask("Generate string", task)
    }
}